src/Dapr.Common/DaprGenericClientBuilder.cs

```diff
@@ -0,0 +1,226 @@
+﻿// ------------------------------------------------------------------------
+// Copyright 2024 The Dapr Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//     http://www.apache.org/licenses/LICENSE-2.0
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// ------------------------------------------------------------------------
+
+namespace Dapr.Common;
+
+using System;
+using System.Net.Http;
+using System.Text.Json;
+using Grpc.Net.Client;
+
+/// <summary>
+/// Builder for building a generic Dapr client.
+/// </summary>
+public abstract class DaprGenericClientBuilder<TClientBuilder> where TClientBuilder : class
+{
+    /// <summary>
+    /// Initializes a new instance of the <see cref="DaprGenericClientBuilder{TClientBuilder}"/> class.
+    /// </summary>
+    protected DaprGenericClientBuilder()
+    {
+        this.GrpcEndpoint = DaprDefaults.GetDefaultGrpcEndpoint();
+        this.HttpEndpoint = DaprDefaults.GetDefaultHttpEndpoint();
+
+        this.GrpcChannelOptions = new GrpcChannelOptions()
+        { 
+            // The gRPC client doesn't throw the right exception for cancellation
+            // by default, this switches that behavior on.
+            ThrowOperationCanceledOnCancellation = true,
+        };
+
+        this.JsonSerializerOptions = new JsonSerializerOptions(JsonSerializerDefaults.Web);
+        this.DaprApiToken = DaprDefaults.GetDefaultDaprApiToken();
+    }
+
+    /// <summary>
+    /// Property exposed for testing purposes.
+    /// </summary>
+    public string GrpcEndpoint { get; private set; }
```

Comment:
With `InternalsVisibleToAttribute`, could these still be `internal`?  Unless users have a specific need, we shouldn't increase the surface area of our public API.

Reply:
Yep